### PR TITLE
claud backend

### PR DESCRIPTION
**Pull request by craterink**
Implement the option to use Claud backend for simple prompts.

### CODE REVIEW

File: gpt_json/gpt.py [hunk 1]
```diff
@@ -69,11 +76,11 @@ def __init__(
         # For messages that are relatively deterministic
         temperature=0.2,
         timeout=60,
-        openai_max_retries=3,
+        api_max_retries=3,
```
Comment (piercefreeman) [inline gpt_json/gpt.py:79]: nit: Let's keep backwards compatibility here. Can you add a new `api_max_retries` parameter to __init__ and flag via a logging.warn that `openai_max_retries` is deprecated?

File: gpt_json/gpt.py [hunk 2]
```diff
@@ -112,9 +119,14 @@ def __init__(
                 self.max_tokens = 8192 - auto_trim_response_overhead
             elif "gpt-3.5" in self.model:
                 self.max_tokens = 4096 - auto_trim_response_overhead
+            elif "claud-v1" in self.model:
+                # todo confirm this number
```
Comment (piercefreeman) [inline gpt_json/gpt.py:123]: Seems like at least in their release notes they cite 9k as the previous limit? https://www.anthropic.com/index/100k-context-windows

File: gpt_json/gpt.py [hunk 3]
```diff
@@ -302,19 +331,46 @@ async def submit_request(
...
+            GPTModelVersion.GPT_3_5,
+            GPTModelVersion.GPT_4,
+        ]:
             optional_parameters["max_tokens"] = max_response_tokens
+        elif self.model in [
+            GPTModelVersion.CLAUD,
+            GPTModelVersion.CLAUD_100K,
+        ]:
+            # need to specify a default for CLAUD models.
+            # using OpenAI's default of 16 tokens: https://platform.openai.com/docs/api-reference/completions/create#completions/create-max_tokens
```
Comment (piercefreeman) [inline gpt_json/gpt.py:344]: Is this a magic number; any empirical reason to choose 16?

**Review by piercefreeman — COMMENTED** [review]
High level: We're redefining the arrays a lot to separate logic for GPT and Claud (ie. if self.model in `GPTModelVersion.GPT_3_5, GPTModelVersion.GPT_4`). I see this growing quickly in the future as more architectures / versions come online.

I'd suggest we create a new class `ClaudModelVersion` enum that is comparable to `GPTModelVersion`. Both can inherit from a new superclass `ModelVersion` - which can be passed as the generic model class wherever we need to determine a model version. Then we can check arbitrary GPT vs. Claud status by just checking the membership of the current enum for the given subclass. Does that sound reasonable?

File: gpt_json/gpt.py [hunk 4]
```diff
@@ -302,19 +331,46 @@ async def submit_request(
...
+                **optional_parameters,
+                **self.api_arguments,
+            )
+        elif self.model in [GPTModelVersion.CLAUD, GPTModelVersion.CLAUD_100K]:
+            # format using Claud-mandated user prompt template
+            # see more: https://console.anthropic.com/docs/api/reference#-v1-complete
+            formatted_claud_prompt = (
+                f"{anthropic.HUMAN_PROMPT} {messages[0].content}{anthropic.AI_PROMPT}"
+            )
+            c = anthropic.Client(self.api_key)
```
Comment (piercefreeman) [inline gpt_json/gpt.py:366]: I don't see these being added to any of the pyproject.toml / requirement files. Can we add a custom installable extra that only pulls in the anthropic dependencies is users specifically install with `gpt-json[anthropic]`?

File: examples/claud_example.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+Respond with the following JSON schema:
+
+{json_schema}
+
+Text: I love this product. It's the best thing ever!
+"""
+
+
+async def runner():
+    gpt_json = GPTJSON[SentimentSchema](API_KEY, model=GPTModelVersion.CLAUD)
```
Comment (piercefreeman) [inline examples/claud_example.py:32]: Can you stub out what the expected output format of the anthropic API is, and write a bit of coverage for this in unit tests?

File: gpt_json/gpt.py [hunk 6]
```diff
@@ -302,19 +331,46 @@ async def submit_request(
...
+            return await c.acompletion(
+                prompt=formatted_claud_prompt,
+                stop_sequences=[anthropic.HUMAN_PROMPT],
+                model="claude-v1",
```
Comment (piercefreeman) [inline gpt_json/gpt.py:370]: Should this be hard-coded, or tie in with the model version so it can do `claud-v1` vs. `claud-v1-100k`?

**Comment by dgonier** [comment]
Nice! Thanks guys!

File: examples/claud_example.py (hunk 5 above)
Comment (dgonier) [reply examples/claud_example.py:32]: Sure. I can do that later this week

File: gpt_json/gpt.py (hunk 3 above)
Comment (craterink) [reply gpt_json/gpt.py:344]: no empirical, just consistency with OAI's api

File: gpt_json/gpt.py (hunk 4 above)
Comment (craterink) [reply gpt_json/gpt.py:368]: the anthropic package is fairly small (24K), just adding it to default package installs for now unless you feel strongly it should be an extra

**Comment by craterink** [comment]
> I'd suggest we create a new class `ClaudModelVersion` enum 

I tried this but using a separate ModelProvider enum ended up feeling cleaner. wdyt?

**Pull request closed without merging**